### PR TITLE
HLSL: Implement GroupOperation(Inclusive/Exclusive)Scan.

### PR DESCRIPTION
**Pull request by RandomShaper**

### CODE REVIEW

File: spirv_hlsl.cpp [hunk 1]
```diff
@@ -4728,21 +4728,33 @@ void CompilerHLSL::emit_subgroup_op(const Instruction &i)
...
+			}
+		} else if (operation == GroupOperationExclusiveScan) {
+			auto expr = join("countbits(", to_enclosed_expression(ops[4]), " & gl_SubgroupLtMask)");
+			emit_op(result_type, id, expr, forward);
+			if (!active_input_builtins.get(BuiltInSubgroupLtMask))
+			{
+				active_input_builtins.set(BuiltInSubgroupLtMask);
+				force_recompile_guarantee_forward_progress();
+			}
+		} else
```
Comment (akien-mga) [inline spirv_hlsl.cpp:4757]: From a quick look in that file upstream seems to favor Allman style:
```suggestion
		else if (operation == GroupOperationInclusiveScan)
		{
			auto expr = join("countbits(", to_enclosed_expression(ops[4]), " & gl_SubgroupLeMask)");
			emit_op(result_type, id, expr, forward);
			if (!active_input_builtins.get(BuiltInSubgroupLeMask))
			{
				active_input_builtins.set(BuiltInSubgroupLeMask);
				force_recompile_guarantee_forward_progress();
			}
		}
		else if (operation == GroupOperationExclusiveScan)
		{
			auto expr = join("countbits(", to_enclosed_expression(ops[4]), " & gl_SubgroupLtMask)");
			emit_op(result_type, id, expr, forward);
			if (!active_input_builtins.get(BuiltInSubgroupLtMask))
			{
				active_input_builtins.set(BuiltInSubgroupLtMask);
				force_recompile_guarantee_forward_progress();
			}
		}
		else
```

Comment (RandomShaper) [reply spirv_hlsl.cpp]: Done!

File: spirv_hlsl.cpp [hunk 2]
```diff
@@ -4739,9 +4739,25 @@ void CompilerHLSL::emit_subgroup_op(const Instruction &i)
 			inherit_expression_dependencies(id, ops[4]);
 		}
 		else if (operation == GroupOperationInclusiveScan)
-			SPIRV_CROSS_THROW("Cannot trivially implement BallotBitCount Inclusive Scan in HLSL.");
+		{
+			auto expr = join("countbits(", to_enclosed_expression(ops[4]), " & gl_SubgroupLeMask)");
```
Comment (HansKristian-Work) [inline spirv_hlsl.cpp:4743]: This is questionable when wave size is > 32. I think you'd need to sum the components like for reduce.

Comment (HansKristian-Work) [reply spirv_hlsl.cpp:4743]: Also, I'm missing tests here.

Comment (RandomShaper) [reply spirv_hlsl.cpp]: Changes done.

**Review by HansKristian-Work — APPROVED** [review]
LGTM.

**Pull request merged**